src/function.ts

```diff
@@ -61,28 +112,347 @@ export class Function<P, O> implements Integration {
+  }
+}
+
+export class ImportedFunction<P, O> extends FunctionBase<P, O> {
+  constructor(func: aws_lambda.IFunction) {
+    return super(func) as unknown as ImportedFunction<P, O>;
+  }
+}
+
+type ConditionalFunction<P, O> = P extends undefined
+  ? (payload?: P) => O
+  : (payload: P) => O;
+
+/**
+ * A special lambda code wrapper that serializes whatever closure it is given.
+ *
+ * Caveat: Relies on async functions which may not finish when using CDK's app.synth()
+ *
+ * Ensure the generate function's promise is completed using something like Lambda.promises and the `asyncSynth` function.
+ *
+ * Use:
+ * * Initialize the {@link CallbackLambdaCode} `const code = new CallbackLambdaCode()`
+ * * First bind the code to a Function `new aws_lambda.Function(..., { code })`
+ * * Then call generate `const promise = code.generate(integrations)`
+ * *
+ */
+export class CallbackLambdaCode extends aws_lambda.Code {
+  private scope: Construct | undefined = undefined;
+
+  constructor(
+    private func: (preWarmContext: NativePreWarmContext) => AnyFunction
+  ) {
+    super();
+  }
+
+  public bind(scope: Construct): aws_lambda.CodeConfig {
+    this.scope = scope;
+    // Lets give the function something lightweight while we process the closure.
+    return aws_lambda.Code.fromInline(
+      "If you are seeing this in your lambda code, ensure generate is called and then consult the README."
+    ).bind(scope);
+  }
+
+  /**
+   * Thanks to cloudy for the help getting this to work.
+   * https://github.com/skyrpex/cloudy/blob/main/packages/cdk/src/aws-lambda/callback-function.ts#L518-L540
+   * https://twitter.com/samgoodwin89/status/1516887131108438016?s=20&t=7GRGOQ1Bp0h_cPsJgFk3Ww
+   */
+  public async generate(
+    integrationPrewarms: NativeIntegration<AnyFunction>["preWarm"][]
+  ) {
+    if (!this.scope) {
+      throw Error("Must first be bound to a Construct using .bind().");
+    }
+    const scope = this.scope;
+    let tokens: string[] = [];
+    const preWarmContext = new NativePreWarmContext();
+    const func = this.func(preWarmContext);
+
+    /* istanbul ignore next */
+    const result = await runtime.serializeFunction(
+      // factory function allows us to prewarm the clients and other context.
+      () => {
+        /* istanbul ignore next */
+        integrationPrewarms.forEach((i) => i?.(preWarmContext));
+        return func;
+      },
+      {
+        isFactoryFunction: true,
+        serialize: (obj) => {
+          if (typeof obj === "string") {
+            const reversed =
+              Tokenization.reverse(obj, { failConcat: false }) ??
+              Tokenization.reverseString(obj).tokens;
+            if (!Array.isArray(reversed) || reversed.length > 0) {
+              if (Array.isArray(reversed)) {
+                tokens = [...tokens, ...reversed.map((s) => s.toString())];
+              } else {
+                tokens = [...tokens, reversed.toString()];
+              }
+            }
```

Comment:
What's this do? Why do we need to reverse things?

Reply:
reverse ~= `de-encode`, and yeah, it helps us find real tokens.